File: apps/marquee/quorum-sense/android/app/src/main/java/se/reflective/quorum/capture/FieldSignalDraft.kt
package se.reflective.quorum.capture

import uniffi.quorum_ffi.AppendEventType
import uniffi.quorum_ffi.ConsentState
import uniffi.quorum_ffi.SignalModality
import uniffi.quorum_ffi.SyncState

// SignalModality, ConsentState, AppendEventType and SyncState are no longer
// declared here: UniFFI now generates them (uniffi/quorum_ffi/quorum_ffi.kt)
// straight from the UDL contract, so the wire enum *is* the domain enum — there
// are no String wire values to parse and an unknown case is unrepresentable.
// These extensions add only the host-side affordance the bindings don't emit:
// the human-facing label. `entries` (the picker's case list) and structural
// equality come from the generated `enum class` for free.

val SignalModality.label: String
    get() = when (this) {
        SignalModality.TEXT -> "Text"
        SignalModality.VOICE_TRANSCRIPT -> "Voice"
        SignalModality.IMAGE_OCR_TEXT -> "Photo OCR"
    }

val ConsentState.label: String
    get() = when (this) {
        ConsentState.PENDING -> "Pending"
        ConsentState.CONSENTED -> "Consented"
    }

val AppendEventType.label: String
    get() = when (this) {
        AppendEventType.SIGNAL_DRAFT_CONSENTED -> "Signal draft consented"
    }

val SyncState.label: String
    get() = when (this) {
        SyncState.QUEUED_FOR_SYNC -> "Queued for sync"
    }

/** A model confidence score constrained to `0f..1f`, mirroring Rust `Confidence`. */
@JvmInline
value class Confidence private constructor(val value: Float) {
    companion object {
        /** Validated construction from untrusted input (e.g. core output). */
        fun of(value: Float): Confidence? =
            if (value.isFinite() && value in 0f..1f) Confidence(value) else null

        /** Trusted construction for compile-time-known-valid literals (previews/tests). */
        fun literal(value: Float): Confidence = Confidence(value)
    }
}

data class FieldSignalDraft(
    val workflowId: String,
    val draftId: String,
    val inquiryThreadId: String,
    val modality: SignalModality,
    val rawCapture: String,
    val summary: String,
    val latentNeed: String,
    val contradiction: String,
    val confidence: Confidence,
    val consentState: ConsentState,
)

data class QuorumAppendEvent(
    val workflowId: String,
    val eventType: AppendEventType,
    val draftId: String,
    val inquiryThreadId: String,
    val syncState: SyncState,
)

File: apps/marquee/quorum-sense/ios/Capture/FieldSignalDraft.swift
import Foundation

public enum SignalModality: String, CaseIterable, Identifiable, Sendable {
    case text
    case voiceTranscript = "voice_transcript"
    case imageOcrText = "image_ocr_text"

    public var id: String { rawValue }

    public var label: String {
        switch self {
        case .text: "Text"
        case .voiceTranscript: "Voice"
        case .imageOcrText: "Photo OCR"
        }
    }
}

/// Whether a captured signal has cleared consent for sync. Mirrors the Rust
/// `ConsentState`; raw values are the wire contract from the core.
public enum ConsentState: String, Sendable {
    case pending
    case consented

    public var label: String {
        switch self {
        case .pending: "Pending"
        case .consented: "Consented"
        }
    }
}

/// The kind of event emitted when a draft is appended. Mirrors Rust `AppendEventType`.
public enum AppendEventType: String, Sendable {
    case signalDraftConsented = "SignalDraftConsented"

    public var label: String {
        switch self {
        case .signalDraftConsented: "Signal draft consented"
        }
    }
}

/// Where an appended event sits in the sync pipeline. Mirrors Rust `SyncState`.
public enum SyncState: String, Sendable {
    case queuedForSync = "queued_for_sync"

    public var label: String {
        switch self {
        case .queuedForSync: "Queued for sync"
        }
    }
}

/// A model confidence score constrained to `0...1`, mirroring Rust `Confidence`.
public struct Confidence: Equatable, Sendable {
    public let value: Float

    /// Validated construction from untrusted input (e.g. core output).
    public init?(_ value: Float) {
        guard value.isFinite, (0...1).contains(value) else { return nil }
        self.value = value
    }

    /// Trusted construction for compile-time-known-valid literals (previews/tests),
    /// the Swift analogue of the same-module direct construction Rust uses.
    init(literal value: Float) {
        self.value = value
    }
}

public struct FieldSignalDraft: Equatable, Identifiable, Sendable {
    public let workflowId: String
    public let draftId: String
    public let inquiryThreadId: String
    public let modality: SignalModality
    public let rawCapture: String
    public let summary: String
    public let latentNeed: String
    public let contradiction: String
    public let confidence: Confidence
    public let consentState: ConsentState

    public var id: String { draftId }
}

public struct QuorumAppendEvent: Equatable, Sendable {
    public let workflowId: String
    public let eventType: AppendEventType
    public let draftId: String
    public let inquiryThreadId: String
    public let syncState: SyncState
}
